### ISSUE TITLE
Preventing losing events when downsampling

### ISSUE THREAD

**Issue by skjerns** [mne-tools/mne-bids-pipeline#1028]
I have the following problem

Our trigger channel sometimes takes a while to "ramp up", producing events in the STI101 channel (1000 Hz) like

`0, 0, 0, 4, 5, 5, 5, 5, 5, 0, 0, 0`

when I downsample these data using `mne-bids-pipeline` to 100 Hz, the resulting trigger channel values are `0, 4, 0`, such that the event is interpreted as a 4 instead of a 5.

Usually I would avoid this by extracting events beforehand with appropriate `min_duration` and `shortest_event` and resampling jointly (`raw, events = raw.resample(events=events)`). However, when using mne-bids-pipeline I don't have this option.

Is there any way to avoid this without relying on the `_events.json` sidecar?

**Comment by hoechenberger** [comment]
Hi!

> The events are correctly inserted as annotations into the resampled pipeline data. However, converting back from annotations to events via `mne.events_from_annotations(raw)` will scramble the trigger code values, _unless_ manually specified.

I don't really understand what you're saying here, could you please provide an example? Doesn't need to be an MWE, just some illustration of what happens vs what you would expect. This will also help us triage whether this is a Pipeline or an MNE-Python problem. Thank you!

**Comment by skjerns** [comment]
> I don't really understand what you're saying here, could you please provide an example? Doesn't need to be an MWE, just some illustration of what happens vs what you would expect. This will also help us triage whether this is a Pipeline or an MNE-Python problem. Are you trying to "eject" your data from the Pipeline? Or where does your issue occur? Thank you!

Sure! I use various trigger codes that do not appear in order, e.g. a list of events extrated from `mne.find_events` would read `4, 99, 2, 4, 234`. During the pipeline, these events are detected (before resampling) and are added as annotations to the `raw` object, with the use of the `events.tsv` (which got created by `write_raw_bids(raw=raw, .., events=events,`). Then later, if I want to convert from annotations back to events, the event codes would be simply enumerated in the order that they appeared, e.g. `1,2,3,1,4`. However, as triggers are not in the same order for each participant, the resulting events code would be scrambled across participants.

I am trying to use MNE to preprocess my data, without epoching. I.e. running ICA, downsampling, filtering etc, and then having the `clean_raw.fif` as input for all further processing :)

**Comment by hoechenberger** [comment]
Ah okay, so you're basically running all pre-processing steps to get the "clean" raw data, and that's when you stop using the pipeline and want to proceed working with the data in your own manual code / pipeline, did I get this right? 🙂

**Comment by skjerns** [comment]
> Ah okay, so you're basically running all pre-processing steps to get the "clean" raw data, and that's when you stop using the pipeline and want to proceed working with the data in your own manual code / pipeline, did I get this right? 🙂

exactly! our final analysis is a bit more intricate than what `mne-bids-pipeline` offers

**Comment by hoechenberger** [comment]
Thanks! Could you share what the annotations on that file look like?

**Comment by skjerns** [comment]
sure! [annotation.zip](https://github.com/user-attachments/files/17899972/annotation.zip)

I think this is intended behaviour in MNE, as far as I can tell, the original trigger values are not saved when running `mne.annotations_from_events` and when running the reverse `mne.events_from_annotations`, the resulting values are just enumerations?

**Comment by hoechenberger** [comment]
You can pass the event_id to control which IDs will be used :)
